File: samples/bot-suggested-actions/csharp/SuggestedActions/Bots/SuggestedActionsBot.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;
using Microsoft.Extensions.Configuration;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

namespace Microsoft.BotBuilderSamples
{
    /// <summary>
    /// This bot will respond to the user's input with suggested actions.
    /// Suggested actions enable your bot to present buttons that the user
    /// can tap to provide input.
    /// </summary>
    public class SuggestedActionsBot : ActivityHandler
    {
        private readonly IConfiguration _configuration;

        public SuggestedActionsBot(IConfiguration config)
        {
            _configuration = config;
        }

        /// <summary>
        /// Invoked when members are added to the conversation.
        /// Sends a welcome message to the user and tells them what actions they may perform to use this bot.
        /// </summary>
        /// <param name="membersAdded">List of members added to the conversation.</param>
        /// <param name="turnContext">Context object containing information for a single turn of conversation with a user.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        protected override async Task OnMembersAddedAsync(IList<ChannelAccount> membersAdded, ITurnContext<IConversationUpdateActivity> turnContext, CancellationToken cancellationToken)
        {
            await SendWelcomeMessageAsync(turnContext, cancellationToken);
        }

        /// <summary>
        /// Invoked when a message activity is received from the user.
        /// Processes the user's input and responds with the appropriate message and suggested actions.
        /// </summary>
        /// <param name="turnContext">Context object containing information for a single turn of conversation with a user.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
        {
            var text = turnContext.Activity.Text.ToLowerInvariant();
            var responseText = ProcessInput(text);

            await turnContext.SendActivityAsync(responseText, cancellationToken: cancellationToken);
            await SendSuggestedActionsAsync(turnContext, cancellationToken);
        }

        /// <summary>
        /// Sends a welcome message to the user and tells them what actions they may perform to use this bot.
        /// </summary>
        /// <param name="turnContext">Context object containing information for a single turn of conversation with a user.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        private async Task SendWelcomeMessageAsync(ITurnContext turnContext, CancellationToken cancellationToken)
        {
            foreach (var member in turnContext.Activity.MembersAdded)
            {
                if (member.Id != turnContext.Activity.Recipient.Id)
                {
                    await turnContext.SendActivityAsync(
                        "Welcome to the suggested actions bot. This bot will introduce you to suggested actions. Please answer the question:",
                        cancellationToken: cancellationToken);
                    await SendSuggestedActionsAsync(turnContext, cancellationToken);
                }
            }
        }

        /// <summary>
        /// Processes the user's input and returns the appropriate response text.
        /// </summary>
        /// <param name="text">The user's input text.</param>
        /// <returns>The response text based on the user's input.</returns>
        private static string ProcessInput(string text)
        {
            const string colorText = "How can I assist you today?";
            return text switch
            {
                "hello" => $"Hello, {colorText}",
                "welcome" => $"Welcome, {colorText}",
                _ => "Please select one action",
            };
        }

        /// <summary>
        /// Creates and sends an activity with suggested actions to the user.
        /// When the user clicks one of the buttons, the text value from the "CardAction" will be
        /// displayed in the channel just as if the user entered the text.
        /// </summary>
        /// <param name="turnContext">Context object containing information for a single turn of conversation with a user.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        private async Task SendSuggestedActionsAsync(ITurnContext turnContext, CancellationToken cancellationToken)
        {
            var reply = MessageFactory.Text("Choose one of the action from the suggested action?");

            var payload = new
            {
                    type = "Teams.chatMessage",
                    data = new
                    {
                        body = new
                        {
                            additionalData = new { },
                            backingStore = new
                            {
                                returnOnlyChangedValues = false,
                                initializationCompleted = true
                            },
                            content = "<at id=\"0\">SuggestedActionsBot</at>"
                        },
                        mentions = new[]
                        {
                            new
                            {
                                additionalData = new { },
                                backingStore = new
                                {
                                    returnOnlyChangedValues = false,
                                    initializationCompleted = false
                                },
                                id = 0,
                                mentioned = new
                                {
                                    additionalData = new { },
                                    backingStore = new
                                    {
                                        returnOnlyChangedValues = false,
                                        initializationCompleted = false
                                    },
                                    odataType = "#microsoft.graph.chatMessageMentionedIdentitySet",
                                    user = new
                                    {
                                        additionalData = new { },
                                        backingStore = new
                                        {
                                            returnOnlyChangedValues = false,
                                            initializationCompleted = false
                                        },
                                        displayName = "Suggested Actions Bot",
                                        id = "28:" + _configuration["MicrosoftAppId"],
                                    }
                                },
                                mentionText = "Suggested Actions Bot"
                            }
                        },
                        additionalData = new { },
                        backingStore = new
                        {
                            returnOnlyChangedValues = false,
                            initializationCompleted = true
                        }
                    }
            };

            reply.SuggestedActions = new SuggestedActions()
            {
                Actions = new List<CardAction>()
                {
                    new CardAction() { Title = "Hello", Type = ActionTypes.ImBack, Value = "Hello" },
                    new CardAction() { Title = "Welcome", Type = ActionTypes.ImBack, Value = "Welcome" },
                    new CardAction() { Title = "@SuggestedActionsBot", Type = "Action.Compose", Value = payload }
                }
            };

            await turnContext.SendActivityAsync(reply, cancellationToken);
        }
    }
}

File: samples/TeamsSDK/Archived/bot-suggested-actions/python/bots/Suggested_Actions_Bot.py
from botbuilder.core import ActivityHandler, TurnContext, MessageFactory
from botbuilder.schema import SuggestedActions, CardAction, ActionTypes
from config import DefaultConfig

CONFIG = DefaultConfig()


class SuggestedActionsBot(ActivityHandler):
    async def on_members_added_activity(self, members_added, turn_context: TurnContext):
        """
        Handles the event when new members are added to the conversation.
        :param members_added: List of members added to the conversation.
        :param context: The TurnContext object containing information about the turn.
        """
        for member in members_added:
            if member.id != turn_context.activity.recipient.id:
                await turn_context.send_activity(
                    "Welcome to the bot! Let's get started with some suggestions."
                )
                await self._send_suggested_actions(turn_context)

    async def on_message_activity(self, turn_context: TurnContext):
        """
        Handles incoming messages and responds based on user input.
        :param turn_context: The TurnContext object containing information about the turn.
        """
        text = turn_context.activity.text

        match text:
            case "Hello":
                await turn_context.send_activity("Hello! How can I assist you today?")
            case "Welcome":
                await turn_context.send_activity("Welcome! How can I assist you today?")
            case _:
                await turn_context.send_activity(
                    "I didn't understand that. Please choose one of the suggested actions."
                )

        await self._send_suggested_actions(turn_context)

    async def _send_suggested_actions(self, turn_context: TurnContext):
        """
        Sends a message with suggested actions to the user.
        :param turn_context: The TurnContext object containing information about the turn.
        """
        actions = SuggestedActions(
            actions=[
                CardAction(type=ActionTypes.im_back, title="Hello", value="Hello"),
                CardAction(type=ActionTypes.im_back, title="Welcome", value="Welcome"),
                CardAction(
                    type="Action.Compose",
                    title="@SuggestedActionsBot",
                    value={
                        "type": "Teams.chatMessage",
                        "data": {
                            "body": {
                                "additionalData": {},
                                "backingStore": {
                                    "returnOnlyChangedValues": False,
                                    "initializationCompleted": True
                                },
                                "content": "<at id=\"0\">SuggestedActionsBot</at>"
                            },
                            "mentions": [
                                {
                                    "additionalData": {},
                                    "backingStore": {
                                        "returnOnlyChangedValues": False,
                                        "initializationCompleted": False
                                    },
                                    "id": 0,
                                    "mentioned": {
                                        "additionalData": {},
                                        "backingStore": {
                                            "returnOnlyChangedValues": False,
                                            "initializationCompleted": False
                                        },
                                        "odataType": "#microsoft.graph.chatMessageMentionedIdentitySet",
                                        "user": {
                                            "additionalData": {},
                                            "backingStore": {
                                                "returnOnlyChangedValues": False,
                                                "initializationCompleted": False
                                            },
                                            "displayName": "Suggested Actions Bot",
                                            "id": "28:" + CONFIG.APP_ID,
                                        }
                                    },
                                    "mentionText": "Suggested Actions Bot"
                                }
                            ],
                            "additionalData": {},
                            "backingStore": {
                                "returnOnlyChangedValues": False,
                                "initializationCompleted": True
                            }
                        }
                    }
                )
            ]
        )
        reply = MessageFactory.text("Choose one of the action from the suggested actions")
        reply.suggested_actions = actions
        await turn_context.send_activity(reply)